File: Data Structure and Algorithm/Assignment 04/Program 02/prog2.py
# Write a program to merge two already sorted list

class node:
    def __init__(self,data):
        self.data = data
        self.next = None
    
class linkedList:
    def __init__(self):
        self.head = None

    def insert_at_last(self,data):
        new_node = node(data)
        if self.head is None:
            self.head = new_node
            return 
        temp = self.head
        while temp.next:
            temp = temp.next
        temp.next = new_node
        new_node.next = None

    def delete_at_first(self):
        if self.head is not None:
            self.head = self.head.next


    def display(self):
        temp = self.head
        while temp:
            print(temp.data,end="-->")
            temp = temp.next
        print("None")

    

def merge_sorted_list(head1,head2):
        head3 = None
        tail = None
        if head1 is None:
            return head2
        if head2 is None:
            return head1
        
        if head1.data < head2.data:
            head3 = head1
            tail = head1
            head1 = head1.next
        else:
            head3 = head2
            tail = head2
            head2 = head2.next
        
        while head1 and head2:
            if head1.data < head2.data:
                tail.next = head1
                tail = head1
                head1 = head1.next
            else:
                tail.next = head2
                tail = head2
                head2 = head2.next

        # Attach remaining part
        if head1:
            tail.next = head1
        else:
            tail.next = head2

        return head3

if __name__ == "__main__":
    l1 = linkedList()
    l2 = linkedList()
    for val in [1, 3, 5]:
        l1.insert_at_last(val)
    for val in [2, 4, 6]:
        l2.insert_at_last(val)

    merged_head = merge_sorted_list(l1.head, l2.head)

    merged_list = linkedList()
    merged_list.head = merged_head
    merged_list.display()

File: Data Structure and Algorithm/Assignment 04/Program 02/prog2.c
// Write a C program to merge two already sorted list.  
// Mention the necessary algorithm.


#include<stdio.h>
#include<stdlib.h>

typedef struct Node{
    int data;
    struct Node * next;
}node;

node *createNode_at_last(){
    node *nw, *p, *head = NULL;
    int ch = 1,data;
    while (ch!=0)
    {
        printf("Enter the data: ");
        scanf("%d",&data);

        nw = (node*)malloc(sizeof(node));
        nw->data = data;
        nw->next = NULL;
        if (head == NULL)
            head = nw;
        else{
            p = head;
            while (p->next!= NULL)
                p = p->next;
            p->next = nw;
            nw->next = NULL;
        }

        printf("Do you want to input:");
        scanf("%d",&ch);
    }
    return head;
}

node *deleteNode_at_first(node *head){
    node *p =head;
    head = head->next;
    free(p);
    return head;
}

void free_list(node* head) {
    node* temp;
    while (head != NULL) {
        temp = head;
        head = head->next;
        free(temp);
    }
}

void display(node *head){
    node *temp = head;
    while(temp!= NULL){
        printf("%d ->",temp->data);
        temp = temp->next;
    }
    printf("NULL\n");
}

node *merge_linked_list(node *head1, node *head2){
    node *head3 = NULL, *tail;
    //base case
    if(head1 ==NULL) return head2;
    if(head2 == NULL) return head1;

    //Initialize the pointer head3 with either of the first elements of the sorted linked list
    if(head1->data < head2->data){
        head3 = tail = head1;
        head1 = head1->next;
    }
    else{
        head3 = tail = head2;
        head2 = head2->next;
    }

    //General case
    while(head1 != NULL && head2 != NULL){
        if(head1->data < head2->data){
            tail->next = head1;
            tail = head1;
            head1= head1->next;
        }
        else{
            tail->next = head2;
            tail = head2;
            head2= head2->next;
        }
    }

    //Adding the remianung part
    if(head1!= NULL)
        tail->next = head1;
    else{
        tail->next= head2;
    }
    return head3;
}

int main(){
    node *head1, *head2,*head3;
    printf("Enter the two sorted linked list:\n");
    head1 = createNode_at_last();
    head2 = createNode_at_last();
    head3 = merge_linked_list(head1,head2);
    display(head3);
    free_list(head1);
    free_list(head2);
    free_list(head3);
}